packages/flutter/lib/src/material/material.dart

```diff
@@ -410,6 +433,30 @@ abstract class InkFeature {
       onRemoved();
   }
 
+  /// Called when the user input that triggered this feature's appearance was confirmed.
+  ///
+  /// Typically causes the ink to propagate faster across the material. By default this
+  /// method does nothing.
+  void confirm() {
+  }
+
+  /// Called when the user input that triggered this feature's appearance was canceled.
+  ///
+  /// Typically causes the ink to gradually disappear. By default this method does
+  /// nothing.
+  void cancel() {
+  }
+
+  /// The ink's color.
+  Color get color => _color;
+  Color _color;
+  set color(Color value) {
+    if (value == _color)
+      return;
+    _color = value;
+    controller.markNeedsPaint();
+  }
+
```

Comment:
i think we should have an abstract class between InkFeature and InkSplash/InkRipple(/InkHighlight?) that hosts these three.

Reply:
I added an abstract class called InteractiveInkFeature